ultimarc/devices/ultrastik.py

```diff
@@ -0,0 +1,177 @@
+                                payload, USTIK_PRE_MESG_LENGTH,
+                                request_type=USBRequestType.REQUEST_TYPE_VENDOR,
+                                recipient=USBRequestRecipient.RECIPIENT_OTHER)
+            resp_3 = self.read(USBRequestCode.ULTRASTIK_EA, 0x0, USTIK_PRE_INTERFACE, ct.c_void_p(), 0x0,
+                                request_type=USBRequestType.REQUEST_TYPE_VENDOR,
+                                recipient=USBRequestRecipient.RECIPIENT_OTHER)
+            if not resp_2 or not resp_3:
+                break
+
+        if resp_2 and resp_3:
+            resp_4 = self.write(USBRequestCode.ULTRASTIK_E9, 0x0, USTIK_PRE_INTERFACE, ct.c_void_p(), 0x0,
+                            request_type=USBRequestType.REQUEST_TYPE_VENDOR,
+                            recipient=USBRequestRecipient.RECIPIENT_OTHER)
+
+        return not False in [resp_1, resp_2, resp_3, resp_4]
+
+
+class UltraStikDevice(USBDeviceHandle):
+    """
+    Manage an UltraStik 360 Joystick (2015 or newer)
+    """
+    class_id = 'ultrastik'  # Used to match/filter devices.
+    class_descr = _('UltraStik 360 Joystick')
+    interface = USTIK_INTERFACE
+
+    def set_controller_id(self, new_controller_id: int) -> bool:
+        """
+        Set the Ultrastik controller ID
+        :param new_controller_id: New control ID to set, must be an integer between 1 and 4.
+        :return: True if successful, otherwise false.
+        """
+        # Calculate new device ID.
+        new_id = USTIK_CONFIG_BASE + (new_controller_id - 1)
+        # Setup an array with zero for all byte values.
+        data = (ct.c_ubyte * USTIK_MESG_LENGTH)(0)
+        data[0] = new_id
+
+        resp = self.write(USBRequestCode.SET_CONFIGURATION, 0x0, USTIK_INTERFACE, data, USTIK_MESG_LENGTH,
+                          request_type=USBRequestType.REQUEST_TYPE_CLASS,
+                          recipient=USBRequestRecipient.RECIPIENT_INTERFACE)
+
+        return resp
+
+    def set_config(self, config_file: str) -> bool:
+        """ Set the joystick configuration """
+
+        config = JSONObject(self.validate_config_base(config_file, USTIK_RESOURCE_TYPES))
+
+        data = UltraStikStruct()
+        data.keepAnalog = 0x11 if config.keepAnalog else 0x11
+        data.mapSize = config.mapSize
+        data.restrictor = 0x09 if config.restrictor else 0x10
+        data.borders = (ct.c_uint8 * 8)(*config.borders)
+        data.map = (ct.c_uint8 * 81)(*[DIRECTION_MAP[v] for v in config.map])
+        data.flash = 0x00
+        data.reserved = (ct.c_uint8 * 3)(*[0, 0, 0])
+
+        payload = (ct.c_uint8 * USTIK_MESG_LENGTH)(0)
+        ct.memmove(ct.addressof(payload), ct.byref(data, USTIK_MESG_LENGTH), USTIK_MESG_LENGTH)
+        _logger.debug(f"  config data : {' '.join('%02X' % b for b in payload)}")
+
+        resp = self.write_raw(USBRequestCode.SET_CONFIGURATION, 0x1, USTIK_INTERFACE, payload, USTIK_MESG_LENGTH,
```

Comment:
This needs to be in a loop to send all the data, currently it only sends one message

Reply:
See reply below explaining that we hard-coded adding 0x0200 to the 'report_id' arg in 'self.write()'.

Reply:
I'll review this...

Reply:
The fix here is to use write_alt with int(0x00) for report_id.  write_alt takes care of the looping though the data.